### ISSUE TITLE
Discrepancy fulldump and indexed records

### ISSUE THREAD

**Issue by dr0i** [hbz/lobid-resources#1649]
> also perhaps @blackwinter knows if there are 400.000 records missing in the nz fulldump comparing to last week.

That's not the case:

- 2023-02-11: 25,264,919
- 2023-02-18: 25,373,318
- 2023-02-25: 25,382,884
- 2023-03-04: 25,418,914

_Originally posted by @blackwinter in https://github.com/hbz/lobid-resources/issues/1646#issuecomment-1459797859_

**Comment by blackwinter** [comment]
I don't have an explanation off the top of my head. Do the numbers match up for the previous runs?

**Comment by dr0i** [comment]
I don't know. I have to investigate old logs. I have problems understand @hagbeck correctly (https://github.com/hbz/lobid-resources/issues/1646): were there first only 400k missing in  test.lobid.org and in alma.lobid.org it was ok, and then one day later also in alma.lobid.org there are missing items, and which one?

**Comment by hagbeck** [comment]
We request our data with http://alma.lobid.org/resources/search?q=*&owner=DE-290&format=bulk and observed on last Monday, that the results from test.lobid.org are significant lesser than the days before. Our plausibility tests shows the problem and stopped. On the next day, the same problem occurs via alma.lobid.org. 

Because our process stops, it is difficult for us to identify examples of the lost records.

**Comment by dr0i** [comment]
http://alma.lobid.org/resources/search?q=*&owner=DE-290 shows `1,660,199` hits. What would you expect  - >2M ?

**Comment by TobiasNx** [comment]
http://alma.lobid.org/resources/search?q=hasItem.heldBy.id%3A%22http%3A%2F%2Flobid.org%2Forganisations%2FDE-290%23%21%22&format=json Other querry has the same result.

**Comment by hagbeck** [comment]
s.a. https://github.com/hbz/lobid-resources/issues/1646#issuecomment-1461827493

**Comment by dr0i** [comment]
It's good that this issue doesn't affect you. I leave it open since it's not fully understood why there is this discrepancy.

**Comment by blackwinter** [comment]
The `docs` count (from "PubHub") does not include duplicates (nor deletions); we're counting unique MMS IDs (MARC 001). What exactly are you counting?

Curiously, Metafacture itself is yielding yet another record count, while `grep` is confirming the latter:

| Date       | PubHub     | Lobid ETL  | Metafacture | Grep       |
|------------|------------|------------|-------------|------------|
| 2023-02-11 | 25,264,919 | 25,003,764 |             |            |
| 2023-02-18 | 25,373,318 | 25,109,346 |             |            |
| 2023-02-25 | 25,382,884 | 25,031,360 |             |            |
| 2023-03-04 | 25,418,914 | 24,983,778 | 25,010,026  | 25,010,026 |

So there's definitely something off here!

<details><summary>Metafacture Flux</summary>

```
"/path/to/alma/prod/baseline.xml.bgzf"
|open-file(decompressConcatenated="true")
|decode-xml
|handle-marcxml
|batch-log(batchSize="1000000")
|literal-to-object(pattern="(?!)")
|encode-json
|write("/dev/null")
;
```

(BTW: Is there a faster way to count records with Flux? I.e., without encoding them to anything.)
</details>

<details><summary>Grep</summary>

```sh
zgrep -Fc '<record>' /path/to/alma/prod/baseline.xml.bgzf
```
</details>

**Comment by dr0i** [comment]
> What exactly are you counting?

Analog to yours:

<details>
Pseudocode

```
metafix.setStrictness(Metafix.Strictness.EXPRESSION);
metafix.setStrictnessHandlesProcessExceptions(true);

opener.setReceiver(new XmlDecoder())//
.setReceiver(xmlElementSplitter)//
.setReceiver(new LiteralToObject())//
.setReceiver(new ObjectThreader<>())//
new StringReader().setReceiver(new XmlDecoder())//
.setReceiver(marcXmlHandler)//
.setReceiver(metafix)
.setReceiver(batchLogger(100000)//
[...]
```
</details>

I trust grep, so `PubHub` is wrong.
Difference between lobid's records and grep is 27k. This is not that much, but not nothing. Worth investigating more?

**Comment by blackwinter** [comment]
> I trust grep, so `PubHub` is wrong.

I'll look into it.

**Comment by dr0i** [comment]
> (BTW: Is there a faster way to count records with Flux? I.e., without encoding them to anything.)

just leave out the `encode-json`

**Comment by blackwinter** [comment]
> just leave out the `encode-json`

Yes, in combination with `literal-to-object` it's not needed (no significant speedup, though). But one of the two has to be present. And `write` is still required as well. I was more thinking along the lines of a "black hole" command (`StreamSink`).

**Comment by dr0i** [comment]
No problem - glad you've found the explanation :+1: .
I close this issue.
